### DISCUSSION TITLE
posts/conc-better_structured_concurrency_for_go/

### DISCUSSION THREAD

**Discussion by giscus** [overstarry/overstarry.github.io#7] [Announcements]
# posts/conc-better_structured_concurrency_for_go/

前言
本文介绍 conc - 一个更好的 go 并发库， sourcegraph 在日常开发中使用go原生并发出现了问题，由此开发了 conc ，相比标准并发代码更优雅，代码更少，下面展示一个例子,可以看出代码减少了许多.
type propagatedPanic struct {
    val   any
    stack []byte
}
 
func main() {
    done := make(chan *propagatedPanic)
    go func() {
        defer func() {
            if v := recover(); v != nil {
                done <- &propagatedPanic{
                    val:   v,
                    stack: debug.Stack(),
                }
            } else {
                done <- nil
            }
        }()
        doSomethingThatMightPanic()
    }()
    if val := <-done; val != nil {
        panic(val)
    }
}

// conc

func main() {
    var wg conc.WaitGroup
    wg.Go(doSomethingThatMightPanic)
    // panics with a nice stacktrace
    wg.Wait()
}
安装
使用以下命令进行安装: go get github.com/sourcegraph/conc

https://jasminides.com/posts/conc-better_structured_concurrency_for_go/

**Comment by lysonw** [comment]
conc.WaitGroup 的例子有误：

func main() {
	var count atomic.Int64
	var wg conc.WaitGroup
	for i := 0; i < 10; i++ {
		wg.Go(func() {
			if i == 7 {                         // count.Load() == 7
				panic("bad thing")
			}
			count.Add(1)
		})
	}
	wg.WaitAndRecover()
	fmt.Println(count.Load())
}

**Reply by overstarry** [reply]
好的，我看看
